src/data/dfinsupp/well_founded.lean

```diff
@@ -0,0 +1,227 @@
+/-
+Copyright (c) 2022 Junyan Xu. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Junyan Xu
+-/
+import data.dfinsupp.lex
+import order.game_add
+import order.antisymmetrization
+import set_theory.ordinal.basic
+
+/-!
+# Well-foundedness of the lexicographic and product orders on `dfinsupp` and `pi`
+
+The primary results are `dfinsupp.lex.well_founded` and the two variants that follow it,
+which essentially say that if `(>)` is a well order on `ι`, `(<)` is well-founded on each
```

Comment:
This sounds wrong.
```suggestion
which essentially say that if `(<)` is a well order on `ι`, `(<)` is well-founded on each
```

Reply:
The direction is correct and I did prove the ι non-finite/well-order versions: see `lex.well_founded'` (with ` well_founded r.swap` assumption) and `lex.well_founded_lt` (with `[well_founded_gt ι]` assumption). The apparent discordance is due to the definition of `pi.lex`: smaller indices `i : ι` is more important.

Reply:
Wait, does that mean `pi.lex` is wrongly defined, or just counterintuitive?

Reply:
It agrees with the intuitive lexicographic order on words (of fixed length); the first letter is the most important, then the second letter, and so on. When restricted to a symmetric group, it also makes the identity permutation a bottom element, which agrees with the convention for Bruhat order etc.

So I think it's not wrongly defined, but just intended to be used for a finite domain, where any linear order is well-founded both way. When the domain is infinite, for example ℕ, we have the infinite sequence (1,0,0,0,...), (0,1,0,0,...), (0,0,1,0,...), (0,0,0,1,...), which is decreasing (violates well-foundedness) with the usual order on ℕ and increasing with the dual order.

Reply:
The only thing I find unnatural about `pi.lex` is the condition `r j i → x j = y j`. It only guarantees that `i` is a `r`-minimal element where `x` and `y` differ, but there may be other minimal elements, so it's sort of ill-behaved unless `r` is trichotomous. For example, consider `r := (<)` in the partial order where 0<1 and 0<2 but 1 and 2 are incomparable, where 0,1,2 indicate the first, second, and third letter in a 3-letter word. Then aab<aba by taking i=1, and aba<aab by taking i=2, so you immediately get a loop. Non-transitivity is also bad: if 0<1<2 but 0≮2, then aab<baa by taking i=0, and baa<aab by taking i=2.

The more natural condition would be `x j ≠ y j → r i j ∨ j = i` which means that we don't care about the values of `x` and `y` at coordinates `j > i` if they already differ at coordinate `i`. Fortunately, this condition is equivalent to `(¬ r i j ∧ j ≠ i) → x j = y j`, which is the current condition but with `r` replaced by `rᶜ ⊓ (≠)`, and that's what I used to state the most general result `lex.well_founded`.

Reply:
If you redefine `pi.lex r s` to be `∀ j, x j = y j ∨ ∃ i, (i = j ∨ r i j) ∧ s (x i) (y i)`, then I think you get `preorder (lex (Π (i : ι), β i))` from `[preorder ι] [Π i, preorder (β i)]`, without the need to assume `[linear_order ι]` as in [pi.lex.partial_order](https://leanprover-community.github.io/mathlib_docs/data/pi/lex.html#pi.lex.partial_order). But maybe that's too complicated to work with ... I wonder if there's a general standard definition in the literature.

Does anyone has any further comments about this PR?

Reply:
I've thought about the `preorder → preorder` problem (also for `finset.colex`) a while back and couldn't reach any satsifying conclusion. Let me just say that from the beginning of your message I thought you would replace `r j i → x j = y j by `¬ r i j → x j = y j`, which is usually quite workable with.

Also you can't have replaced `r` by `rᶜ ⊓ (≠)`, Surely you rather mean `r ⊔ (=)`?

Reply:
There are three separate things: if you want to replace `well_founded_gt ι` by `well_founde_lt ι` [here](https://github.com/leanprover-community/mathlib/pull/16772/files#diff-563bb010b6a025bae8a967f55f6f52525e9b511143e8012ca775c4977f2784e5R155), then you'd like to replace `r` by `r.swap`. If you want to avoid [loop pathologies](https://github.com/leanprover-community/mathlib/pull/16772#discussion_r1011221268) when ι is not a linear order, then you'd like to replace `r` by `r.swapᶜ ⊓ (≠)`. (I want both, so [lex.well_founded](https://github.com/leanprover-community/mathlib/pull/16772/files#diff-563bb010b6a025bae8a967f55f6f52525e9b511143e8012ca775c4977f2784e5R144) uses `rᶜ ⊓ (≠)`.) Finally, if you want to keep transitivity, then you need `i` to depend on `j`.

By `r ⊔ (=)`, do you mean `x j = y j ∨ j = i ∨ r i j`? Note that's equivalent to `¬ r i j ∧ i ≠ j → x j = y j`, which is the effect of replacing `r` by `r.swapᶜ ⊓ (≠)`.